AppKit/CPSegmentedControl.j

```diff
@@ -677,63 +707,39 @@ CPSegmentSwitchTrackingMomentary = 2;
 {
 }
 
-- (void)tileWithChangedSegment:(unsigned)aSegment
+- (void)tileWithChangedSegment:(CPInteger)aSegment
 {
-    if (aSegment >= _segments.length)
-        return;
-
-    var segment = _segments[aSegment],
-        segmentWidth = [segment width],
-        themeState = _themeState.hasThemeState(CPThemeStateDisabled) ? _themeStates[aSegment].and(CPThemeStateDisabled) : _themeStates[aSegment],
-        contentInset = [self valueForThemeAttribute:@"content-inset" inState:themeState],
-        font = [self font];
+    var segmentCount = [self segmentCount];
 
-    if (!segmentWidth)
+    if (segmentCount == 0)
     {
-        if ([segment image] && [segment label])
-            segmentWidth = [[segment label] sizeWithFont:font].width + [[segment image] size].width + contentInset.left + contentInset.right;
-        else if (segment.image)
-            segmentWidth = [[segment image] size].width + contentInset.left + contentInset.right;
-        else if (segment.label)
-            segmentWidth = [[segment label] sizeWithFont:font].width + contentInset.left + contentInset.right;
-        else
-            segmentWidth = 0.0;
-    }
-
-    var delta = segmentWidth - CGRectGetWidth([segment frame]);
-
-    if (!delta)
-    {
-        [self setNeedsLayout];
-        [self setNeedsDisplay:YES];
-
+        [self setFrameSize:CGSizeMake(0, [self valueForThemeAttribute:@"min-size"].height)];
         return;
     }
 
-    // Update control size
-    var frame = [self frame];
-    [self setFrameSize:CGSizeMake(CGRectGetWidth(frame) + delta, CGRectGetHeight(frame))];
+    if (aSegment < 0 || aSegment >= segmentCount)
+        return;
 
-    // Update segment width
-    [segment setWidth:segmentWidth];
-    [segment setFrame:[self frameForSegment:aSegment]];
+    // Invalidate frames for segments on the right. They will be lazily computed by -frameForSegment:.
+    for (var i = aSegment; i < segmentCount; i++)
+        [_segments[i] setFrame:CGRectMakeZero()];
 
-    // Update following segments widths
-    var index = aSegment + 1;
+    [self setFrameSize:[self intrinsicContentSize]];
 
-    for (; index < _segments.length; ++index)
-    {
-        [_segments[index] frame].origin.x += delta;
+    [self setNeedsLayout];
+    [self setNeedsDisplay:YES];
+}
 
-        [self drawSegmentBezel:index highlight:NO];
-        [self drawSegment:index highlight:NO];
-    }
+- (CGSize)intrinsicContentSize
```

Comment:
It's public in cocoa but since this is the only occurence in AppKit and it won't be documented, I will privatize it. Although i think all controls should implement this method because it gives a better separation between metrics and actual layout.